server/data/data-cassandra/src/main/java/org/apache/james/sieve/cassandra/CassandraSieveRepository.java

```diff
@@ -0,0 +1,276 @@
+import org.apache.james.sieverepository.api.ScriptSummary;
+import org.apache.james.sieverepository.api.SieveRepository;
+import org.apache.james.sieverepository.api.exception.DuplicateException;
+import org.apache.james.sieverepository.api.exception.IsActiveException;
+import org.apache.james.sieverepository.api.exception.QuotaExceededException;
+import org.apache.james.sieverepository.api.exception.QuotaNotFoundException;
+import org.apache.james.sieverepository.api.exception.ScriptNotFoundException;
+import org.apache.james.sieverepository.api.exception.StorageException;
+import org.joda.time.DateTime;
+
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+
+public class CassandraSieveRepository implements SieveRepository {
+
+    private final CassandraSieveDAO cassandraSieveDAO;
+
+    @Inject
+    public CassandraSieveRepository(Session session) {
+        this.cassandraSieveDAO = new CassandraSieveDAO(session);
+    }
+
+    @Override
+    public DateTime getActivationDateForActiveScript(String user) throws StorageException, ScriptNotFoundException {
+        return new DateTime(cassandraSieveDAO.getActiveScriptMetadataRow(user)
+            .join()
+            .orElseThrow(ScriptNotFoundException::new)
+            .getDate(CassandraSieveTable.DATE)
+            .getTime());
+    }
+
+    @Override
+    public void haveSpace(String user, String name, long newSize) throws QuotaExceededException, StorageException {
+        throwOnOverQuota(user, spaceThatWillBeUsedByNewScript(user, name, newSize));
+    }
+
+    private void throwOnOverQuota(String user, CompletableFuture<Long> sizeDifference) throws QuotaExceededException, StorageException {
+        CompletableFuture<Optional<Long>> userQuotaFuture = cassandraSieveDAO.getQuota(user);
+        CompletableFuture<Optional<Long>> globalQuotaFuture = cassandraSieveDAO.getQuota();
+        CompletableFuture<Long> spaceUsedFuture = cassandraSieveDAO.spaceUsedBy(user);
+
+        throwOnOverQuotaInternals(sizeDifference.join(),
+            quotaToUse(userQuotaFuture, globalQuotaFuture),
+            spaceUsedFuture.join());
+    }
+
+    public CompletableFuture<Long> spaceThatWillBeUsedByNewScript(String user, String name, long scriptSize) {
+        return cassandraSieveDAO.getScriptMetadataRow(user, name)
+            .thenApply(optional -> optional.map(row -> row.getLong(CassandraSieveTable.SIZE)))
+            .thenApply(sizeOfStoredScript -> scriptSize - sizeOfStoredScript.orElse(0L));
+    }
+
+    private Optional<Long> quotaToUse(CompletableFuture<Optional<Long>> userQuota, CompletableFuture<Optional<Long>> globalQuota) {
+        if (userQuota.join().isPresent()) {
+            return userQuota.join();
+        }
+        return globalQuota.join();
+    }
+
+    private void throwOnOverQuotaInternals(Long spaceUsedByNewScript, Optional<Long> quota, long spaceUsed) throws QuotaExceededException {
```

Comment:
why is this part specific to cassandra repository ? I would rather see a Quota object with some `checkOverQuota()` method, or anything that looks like OOP